router/src/harness/include/mysql/harness/mpsc_queue.h

```diff
@@ -43,7 +43,7 @@ class MPSCQueueDV {
  public:
   using value_type = T;
 
-  MPSCQueueDV()
+  MPSCQueueDV() noexcept
```

Comment:
What does this fix and why is this correct - given "new Node" below?

Reply:
It addresses this error emitted from gcc 4.8 only:
/home/travis/build/percona/percona-server/router/src/harness/src/loader.cc:27:/home/travis/build/percona/percona-server/router/src/harness/include/mysql/harness/mpsc_queue.h: In instantiation of ‘mysql_harness::MPSCQueueDV<T>::MPSCQueueDV() [with T = std::__exception_ptr::exception_ptr]’:/home/travis/build/percona/percona-server/router/src/harness/include/mysql/harness/waiting_queue_adaptor.h:38:7:   required from here/home/travis/build/percona/percona-server/router/src/harness/include/mysql/harness/mpsc_queue.h:111:10: error: use of deleted function ‘std::atomic<_Tp*>::atomic(const std::atomic<_Tp*>&) [with _Tp = mysql_harness::MPSCQueueDV<std::__exception_ptr::exception_ptr>::Node]’   struct Node {          ^In file included from /home/travis/build/percona/percona-server/router/src/harness/include/mysql/harness/mpsc_queue.h:28:0,                 from /home/travis/build/percona/percona-server/router/src/harness/include/mysql/harness/loader.h:626,                 from /home/travis/build/percona/percona-server/router/src/harness/src/loader.cc:27:/usr/include/c++/4.8/atomic:298:7: error: declared here       atomic(const atomic&) = delete;       ^In file included from /home/travis/build/percona/percona-server/router/src/harness/include/mysql/harness/loader.h:626:0,                 from /home/travis/build/percona/percona-server/router/src/harness/src/loader.cc:27:/home/travis/build/percona/percona-server/router/src/harness/include/mysql/harness/mpsc_queue.h:47:69: note: synthesized method ‘mysql_harness::MPSCQueueDV<std::__exception_ptr::exception_ptr>::Node::Node()’ first required here        : head_{new Node}, tail_{head_.load(std::memory_order_relaxed)} {}

I am unsure why this only trips on gcc 4.8.  Stolen from a stackoverflow on this identical question
> std::atomic's default constructor is noexcept, so when you try to instantiate it with some type, that type also needs to be noexcept default constructible so that the exception specifications match.

Reply:
OK, but the 2nd half of my question stands: why is this correct, given "new Node" in constructor body which may throw std::bad_alloc?

Reply:
I have no idea why, only that this fixed the error.  Considering that his was limited to gcc 4.8, I suspect that this is a compiler implementation fail.

Reply:
Just ran into this. Looks like https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60258#c4

Adding `noexcept` doesn't fix the compile error here.